### DISCUSSION TITLE
How to prevent `()` from being inserted after completion?

### DISCUSSION THREAD

**Discussion by avatar-lavventura** [company-mode/company-mode#1218] [Q&A] [answered]
When I apply `company-complete-selection`, it completes variable but sometimes adds `()` at the end of it. Would it be possible to prevent this in all cases?

Example:
[![enter image description here][2]][2]

which completed as: `self.logged_jobs_to_process()`

------

my setup:

```
(add-hook 'after-init-hook 'global-company-mode)
(setq company-auto-commit t)
(setq company-auto-complete t)
(global-set-key (kbd "C-c C-k") 'company-complete)
(setq company-frontends
      '(company-pseudo-tooltip-unless-just-one-frontend
        company-preview-frontend
        company-echo-metadata-frontend))
```

  [1]: https://company-mode.github.io
  [2]: https://i.stack.imgur.com/DG0OK.png

**Comment by dgutov** [comment, marked as answer]
Any post-processing like that is implemented in the backend. It can only be disabled in the backend as well -- and if there is no such option, you can ask the author to add it.

To find out which backend you are using, type `M-x company-diag`.

**Comment by avatar-lavventura** [comment]
`M-x company-diag` returns:

```
Emacs 28.0.50 (x86_64-pc-linux-gnu) of 2021-09-01 on home
Company 0.9.13

company-backends: (company-bbdb company-semantic company-cmake company-capf company-clang company-files
              (company-dabbrev-code company-gtags company-etags company-keywords)
              company-oddmuse company-dabbrev)

Used backend: company-capf
Value of c-a-p-f: (lsp-completion-at-point)
Major mode: python-mode
Prefix: "None"
Completions:
  #("None" 0 4 (lsp-completion-item #s(hash-table size 7 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("label" #("None" 0 4 (match-data (0 4 0 1 1 2 2 3 3 4))) "kind" 14
"sortText" "aNone" "insertText" "None" "data" #s(hash-table size 1 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("doc_uri" "file:///home/alper/ebloc-broker/broker/Driver.py"))
"_emacsStartPoint" 3652)) lsp-sort-text "aNone" lsp-completion-start-point 3652 lsp-completion-markers (3652 #<marker (moves after insertion) at 3656 in Driver.py>) lsp-completion-prefix
"None" match-data (0 4 0 1 1 2 2 3 3 4) face (completions-common-part))) " (Keyword)"
```

Seems like it depends on `lsp-completion-at-point`. So I will have to get contact with `lsp` developers.

**Reply by yugaego** [reply]
> it depends on lsp-completion-at-point

That's right. Moreover, the chances are high the Python **server** you have installed is responsible. See also a somewhat related [Issue in lsp-mode](https://github.com/emacs-lsp/lsp-mode/issues/1033).

**Reply by avatar-lavventura** [reply]
Doing `(setq lsp-enable-snippet nil)` solved the problem I was having.

**Reply by dgutov** [reply]
Very good. I though you wanted to skip inserting `()` but keep inserting non-empty arguments, and that's more complicated.
